modules/gmake/src/proj.lua

```diff
@@ -0,0 +1,638 @@
+		return {
+			proj.header,
+			proj.defaultConfigurations,
+			proj.verbosity,
+			proj.phonies,
+			proj.shellType,
+			proj.defines,
+			proj.includeDirs,
+			proj.targetDir,
+			proj.intermediateDir,
+			proj.targetName,
+			proj.cppFlags,
+			proj.cFlags,
+			proj.cxxFlags,
+			proj.linkFlags,
+			proj.configurations,
+			proj.linkCmd,
+			proj.objects,
+			proj.allRule,
+			proj.targetRule,
+			proj.targetDirRule,
+			proj.objDirRule,
+			proj.cleanRule,
+			proj.prebuildRule,
+			proj.fileRules
+		}
+	end,
+	configurations = function (cfg)
+		return {
+			proj.targetDir,
+			proj.intermediateDir,
+			proj.targetName,
+			proj.defines,
+			proj.cppFlags,
+			proj.cFlags,
+			proj.cxxFlags,
+			proj.includeDirs,
+			proj.linkFlags,
+			proj.configBuildCommands
+		}
+	end
+}
+
+
+proj.HEADER_FILES = array.join(premake.C_HEADER_EXTENSIONS, premake.CXX_HEADER_EXTENSIONS, premake.OBJC_HEADER_EXTENSIONS)
+proj.SOURCE_FILES = array.join(premake.C_SOURCE_EXTENSIONS, premake.CXX_SOURCE_EXTENSIONS, premake.OBJC_SOURCE_EXTENSIONS)
+
+
+---
+-- Local helper function for determining if an object is a project.
+--
+-- @param obj
+--  object to determine if is a project
+-- @returns if the object is a project.  False can typically be interpretted that the object is a configuration.
+---
+local function isProject(obj)
+	return obj.project == nil
+end
+
+
+---
+-- Export the project's `.makefile` file.
+--
+-- @return
+--	True if the target `.makefile` file was updated; false otherwise.
+---
+function proj.export(prj)
+	export.eol('\r\n')
+	export.indentString('\t')
+	premake.export(prj, prj.exportPath, function()
+		premake.callArray(proj.elements.project, prj)
+	end)
+end
+
+
+---
+-- Prints the header at the top of the project file.
+--
+-- @param prj
+--  project to print header for
+---
+function proj.header(prj)
+	wl('# GNU Makefile project file autogenerated by Premake')
+	wl()
+end
+
+
+---
+-- Prints the default configuration information for the project.
+--
+-- @param prj
+--  project to print defaults for
+---
+function proj.defaultConfigurations(prj)
+	wl('ifndef config')
+	export.indent()
+	wl('config=' .. prj.configs[1].name:lower())
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints verbosity flags to the makefile.
+--
+-- @param prj
+--  project to print verbosity flags for
+---
+function proj.verbosity(prj)
+	wl('ifndef verbose')
+	export.indent()
+	wl('SILENT = @')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the shell type to the makefile
+--
+-- @param prj
+--  project to print shell type for
+---
+function proj.shellType(prj)
+	wl('SHELLTYPE := posix')
+	wl('ifeq (.exe,$(findstring .exe,$(ComSpec)))')
+	export.indent()
+	wl('SHELLTYPE := msdos')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the include directories for a project to the makefile.
+--
+-- @param prj
+--  project to print include directories for
+---
+function proj.includeDirs(prj)
+	if isProject(prj) then
```

Comment:
We should make `is…()` part of the DOM, since all exporters are likely to need something like it at some point to distinguish between project vs. file vs. configuration, etc.